File: swift/list.swift
//  Created by 정호영 on 2017. 2. 14..
//  Copyright © 2017년 watple. All rights reserved.

import Foundation

class MyNode <T> {
    var item:T
    var next:MyNode?
    
    init(_ item:T) {
        self.item = item
    }
}

let node = MyNode("hello")
print(node, node.item, node.next ?? "null")


struct MyList <T> {
    var head:MyNode<T>
    var count:Int = 0
    
    init(_ item:T) {
        self.head = MyNode(item)
    }ㅌㅂ
    
    mutating func addFirst(_ item:T) {
        let node = MyNode(item)
        node.next = self.head
        self.head = node
        self.count += 1
    }
    
}

var list = MyList(1)
list.addFirst(2)
list.addFirst(3)
print(list.head.item)
print(list.head.next!.item)
print(list.count)

File: c/list.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Node Node;
struct Node {
	int value;
	Node *next;
};

void addFirst(Node **head, int v) {
	Node *newNode = (Node * )malloc(sizeof(Node));
	newNode->value =  v;
	newNode->next = *head;
	*head = newNode;
}

void printList(Node *head) {
	Node *curr = head;
	while (curr != NULL) {
		printf("%d\n", curr->value);
		curr = curr->next;
	}
}

Node *init(int v) {
	Node *head = (Node *)malloc(sizeof(Node));
	head->value = v;
	head->next = NULL;
	return head;
}

int main() {
	Node *head = init(1);
	addFirst(&head, 2);
	addFirst(&head, 3);
	printList(head);
	return 0;
}
